### ISSUE TITLE
Error while searching for files with a given file extension (macOS)

### ISSUE THREAD

**Issue by victordomingos** [victordomingos/Count-files#31]
```
$ countfiles -fe txt /

Recursively searching for .txt files in /.

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/bin/countfiles", line 11, in <module>
    load_entry_point('countfiles', 'console_scripts', 'countfiles')()
  File "/Users/username/Dropbox/Count-files/countfiles/__main__.py", line 102, in main_flow
    include_hidden=include_hidden)
  File "/Users/username/Dropbox/Count-files/countfiles/utils/word_counter.py", line 84, in get_files_by_extension
    in Path(os.path.expanduser(location)).rglob(f"*.{extension}")
  File "/Users/username/Dropbox/Count-files/countfiles/utils/word_counter.py", line 83, in <listcomp>
    files = sorted([f for f
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pathlib.py", line 1094, in rglob
    for p in selector.select_from(self):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pathlib.py", line 543, in _select_from
    for starting_point in self._iterate_directories(parent_path, is_dir, scandir):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pathlib.py", line 533, in _iterate_directories
    for p in self._iterate_directories(path, is_dir, scandir):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pathlib.py", line 533, in _iterate_directories
    for p in self._iterate_directories(path, is_dir, scandir):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pathlib.py", line 531, in _iterate_directories
    if entry.is_dir() and not entry.is_symlink():
OSError: [Errno 9] Bad file descriptor: '/dev/fd/4'
```

This error does not occur when using the counting feature (`countfiles /`), only when searching for file extensions.

Since `/dev/` is composed of device files, I would suggest to simply skip it on Linux and macOS systems. Any additional thoughts?

**Comment by NataliaBondarenko** [comment]
I had the idea to exclude some folders from the search. The program is initially designed to count files in the developer's projects. And not for processing of system files of OS.
Though on Windows it does not give out errors (```python -m countfiles -fe txt /```).
Our program can be considered safe, because it does not change files. But it's better not to touch the system folders.

**Comment by victordomingos** [comment]
So... It seems that `pathlib.Path.rglob()` does not cope well with the kind of stuff that we can find in `/dev`. But `os.walk()` seems to work fine, without any errors. We could explicitly exclude some folders, and probably it's a good idea to do so, but I am afraid that these kind of special files (or links to them) could eventually show up in other places. Maybe it will be easier to use `os.walk()` after all.

I have just made a few changes and it seems to work. But we have now some additional redundant code that needs to be refactored (like functions that do basically the same, with slightly different options). We can schedule that refactoring for v.1.5, though.

@NataliaBondarenko, can you please take a look and see if it still works properly on Windows?

**Comment by NataliaBondarenko** [comment]
We can use ```def non_recursive_search``` and ```def recursive_search``` in  ```def get_files_by_extension```. And then filter the results(paths) by extension with ```def get_file_extension```.
os.walk() and os.scandir() works correctly.

**Comment by victordomingos** [comment]
> We can use def non_recursive_search and def recursive_search in def get_files_by_extension. And then filter the results(paths) by extension with def get_file_extension.

@NataliaBondarenko, I was thinking the other way around (getting rid of `recursive_search()` and `non_recursive_search()`, and using `get_files_with_extension()` instead. This function can replace  both `recursive_search()` and `non_recursive_search()`, and even `get_files_without_extension()`, by adding an if clause to check for the presence of an extension. The resulting function could be renamed to something simpler, like `get_files()`. I am trying this approach right now. It will probably imply further changes in `word_counter.py` and `__main__.py`.

**Comment by NataliaBondarenko** [comment]
Try it, we need to test different options.
Just update the comments to the function. it's not entirely clear in ```def get_files()```:
```elif not extension:```, extension: str <- this is a required parameter.

**Comment by victordomingos** [comment]
Yes, it’s required, but `not extension` returns False if the content of `extension` is *Falsy*, like for instance an empty string. 

The docstring may need to be updated to show the purpose of that parameter with empty string. An empty string means “all files, with or without extension”. A string with `.` means “files without extension”. 

I had done that but maybe it was lost with some undo at some point...

**Comment by NataliaBondarenko** [comment]
Good. Did you add a new option, ```all the files```? But now the function does not return a list of objects <class 'pathlib.Path'>, but a list of strings for os.walk() and list with os.DirEntry objects  for os.scandir()

**Comment by victordomingos** [comment]
Yeah, need to correct that.

**Comment by victordomingos** [comment]
Just a quick notice to let you know that I submitted an issue for this on the Python Bug Tracker. Seems to be something specific to Mac, probably a bug in the operating system...

https://bugs.python.org/issue33635

**Comment by victordomingos** [comment]
Just a quick notice to let you know that the Python team seems to have finally solved this bug in the just released Python 3.7.1.

https://bugs.python.org/issue33635

This project has already contributed to a better Python world :-)
